### DISCUSSION TITLE
재사용 가능한 구조적 설계

### DISCUSSION THREAD

**Discussion by june-777** [woowa-techcamp-2024/Team1-3K1K#126] [Ideas]
## Context
> **재사용 가능한 구조적 설계**에 대하여 논의하고자 합니다.

현재 저희의 진행방식은 아래와 같습니다.
- 하나의 기능에 대해 다양한 방식으로 고도화하고, 성능 측정 및 비교하여 최종본을 Main Merge

<br>

**예상되는 문제점**은 아래와 같습니다.
- 재사용 가능한 구조가 아니기 때문에, 추 후 통합테스트를 진행할 때, 고도화 방식을 교체해야할 때 등 제약이 있습니다.

<br>

## 1안. 전략 패턴을 차용
> 재사용 가능한 설계의 여러 방안중 전략 패턴을 제안합니다.

AS-IS
- 담당한 기능 구현에 대하여 성능 측정 및 비교하여, 더 적합한 기능을 main에 Merge
- i.g. 메뉴 재고수 시스템 설계 - 낙관적락 ver, 비관적락 ver, Redis ver
  - 최종본: Redis ver

TO-BE
- 담당한 기능 구현에 대하여 성능 측정 및 비교하고, 전략 패턴을 차용
- i.g. 메뉴 재고수 시스템 설계 - 낙관적락 ver, 비관적락 ver, Redis ver
  - strategy1: 낙관적락
  - strategy2: 비관적락
  - strategy3: Redis

**Comment by Dr-KoKo** [comment]
동일한 기능을 여러 방식으로 구현해야 하는 상황입니다.

각 방식마다 새로운 브랜치에서 작업하게 되면 모듈을 갈아끼워가면서 성능 테스트하는 데에 제약이 있습니다.
따라서 준기님의 의견대로 전략패턴 (또는 다양한 방식을 이용해서) 모듈을 쉽게 갈아끼울 수 있는 구조로 발전해야할 것 같아요.

기능이 완성되면 각각의 버전 별로 release하면 될 것 같습니다.
